### PR TITLE
IF: Implement proposer policy change

### PR DESCRIPTION
**Pull request by heifner**
- Implement proposer policy change design: https://hackmd.io/@9xzIZ7ZaRwegCHgQW02r3Q/BkQ4Oc4Up
- Producer schedule is changed after `set_proposed_producer` in the next, next producer round.
  - For example: Producer A [ b1, b2, .. b12 ], Producer B [ b1, b2, .. b12 ], Producer X [ b1, b2, .. b12 ]
    - If `set_proposed_producer` is called in trx in any blocks of Producer A [ b1, b2, .. b12], the schedule change happens at Producer X [b1]. 
- public controller `pending_producers()` and `proposed_producers` renamed to `_legacy`.
- Added `next_producers()` for use by external enties.

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/block_header_state.cpp [hunk 1]
```diff
@@ -61,13 +62,29 @@ block_header_state block_header_state::next(block_header_state_input& input) con
...
       .previous          = input.parent_id,
       .transaction_mroot = input.transaction_mroot,
       .action_mroot      = input.action_mroot,
-      //.schedule_version = ?, [greg todo]
-      //.new_producers = ?     [greg todo]
+      .schedule_version  = header.schedule_version
    };
 
+   if(!proposer_policies.empty()) {
+      auto it = proposer_policies.begin();
```
Comment (greg7mdp) [inline libraries/chain/block_header_state.cpp:73]: Do we want the earliest one?

File: libraries/chain/block_header_state.cpp [hunk 2]
```diff
@@ -61,13 +62,29 @@ block_header_state block_header_state::next(block_header_state_input& input) con
...
+      if (it->first <= input.timestamp) {
+         result.active_proposer_policy = it->second;
+         result.header.schedule_version = header.schedule_version + 1;
+         result.active_proposer_policy->proposer_schedule.version = result.header.schedule_version;
+         result.proposer_policies = { ++it, proposer_policies.end() };
+      } else {
+         result.proposer_policies = proposer_policies;
```
Comment (greg7mdp) [inline libraries/chain/block_header_state.cpp:80]: should we also have here:
`result.proposer_policy = proposer_policy;`

File: libraries/chain/block_header_state.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/block_header_state.cpp:73]: They are ordered by timestamp. We want the earliest timestamp for our comparison with current (input.timestamp).

File: libraries/chain/block_header_state.cpp (hunk 2 above)
Comment (heifner) [reply libraries/chain/block_header_state.cpp:80]: I could have swore I had that in there. Yes, we want `result=active_proposer_policy = active_proposer_policy`. I can push a commit with that fix.

Comment (heifner) [reply libraries/chain/block_header_state.cpp:80]: pushed

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Implement proposer policy for Faster Finality. Existing proposal policy renamed to reflect they are legacy.
Note: end

**Pull request merged**